src/musig.rs

```diff
@@ -0,0 +1,1286 @@
+            }
+        }
+    }
+
+    /// Obtains the aggregate public key for this [`MusigKeyAggCache`]
+    pub fn agg_pk(&self) -> XOnlyPublicKey {
+        self.1
+    }
+
+    /// Obtains the aggregate public key for this [`MusigKeyAggCache`] as a full [`PublicKey`].
+    ///
+    /// This is only useful if you need the non-xonly public key, in particular for
+    /// plain (non-xonly) tweaking or batch-verifying multiple key aggregations
+    /// (not supported yet).
+    pub fn agg_pk_full(&self) -> PublicKey {
+        unsafe {
+            let mut pk = PublicKey::from(ffi::PublicKey::new());
+            if ffi::secp256k1_musig_pubkey_get(
+                ffi::secp256k1_context_no_precomp,
+                pk.as_mut_c_ptr(),
+                self.as_ptr(),
+            ) == 0
+            {
+                // Returns 0 only if the keys are malformed that never happens in safe rust type system.
+                unreachable!("All the arguments are valid")
+            } else {
+                pk
+            }
+        }
+    }
+
+    /// Apply ordinary "EC" tweaking to a public key in a [`MusigKeyAggCache`].
+    ///
+    /// This is done by adding the generator multiplied with `tweak32` to it. Returns the tweaked [`PublicKey`].
+    /// This is useful for deriving child keys from an aggregate public key via BIP32.
+    /// This function is required if you want to _sign_ for a tweaked aggregate key.
+    ///
+    /// # Arguments:
+    ///
+    /// * `secp` : [`Secp256k1`] context object initialized for verification
+    /// * `tweak`: tweak of type [`SecretKey`] with which to tweak the aggregated key
+    ///
+    /// # Errors:
+    ///
+    /// If resulting public key would be invalid (only when the tweak is the negation of the corresponding
+    /// secret key). For uniformly random 32-byte arrays(for example, in BIP 32 derivation) the chance of
+    /// being invalid is negligible (around 1 in 2^128).
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// # # [cfg(any(test, feature = "rand-std"))] {
+    /// # use secp256k1_zkp::rand::{thread_rng, RngCore};
+    /// # use secp256k1_zkp::{MusigKeyAggCache, Secp256k1, SecretKey, Keypair, PublicKey};
+    /// # let secp = Secp256k1::new();
+    /// # let sk1 = SecretKey::new(&mut thread_rng());
+    /// # let pub_key1 = PublicKey::from_secret_key(&secp, &sk1);
+    /// # let sk2 = SecretKey::new(&mut thread_rng());
+    /// # let pub_key2 = PublicKey::from_secret_key(&secp, &sk2);
+    /// #
+    /// let mut key_agg_cache = MusigKeyAggCache::new(&secp, &[pub_key1, pub_key2]);
+    ///
+    /// let tweak = SecretKey::from_slice(b"Insecure tweak, Don't use this!!").unwrap(); // tweak could be from BIP32
+    /// let _tweaked_key = key_agg_cache.pubkey_ec_tweak_add(&secp, tweak).unwrap();
+    /// # }
+    /// ```
+    pub fn pubkey_ec_tweak_add<C: Verification>(
+        &mut self,
+        secp: &Secp256k1<C>,
+        tweak: SecretKey,
```

Comment:
This should be `Scalar` I believe.

Reply:
Done in https://github.com/rust-bitcoin/rust-secp256k1/pull/716/commits/379f70a152c5be546c678b4179b2ef99c7d955fb